### ISSUE TITLE
Improve cleaning and internal k-fold / splitting with grouped data

### ISSUE THREAD

**Issue by DM-Berger** [stfxecutables/df-analyze#44]
For classification data with a small number (e.g. 5 or less) of target classes, and where the `--grouper` variable is specified, internal k-folds will not in general be able to respect that grouping variable in a large number of cases (https://github.com/scikit-learn/scikit-learn/issues/33085). However, for certain kinds of data (e.g. where the grouper is an identifier), ignoring the grouping can result in extremely biased performance, and likely sub-par tuning and feature selection. 

In these cases, it would probably be preferable to have internal splitting be implemented by [`LeavePGroupsOut`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.LeavePGroupsOut.html), where `P` is chosen so that there are about 3-5 internal splits, for the given grouping variable. If splits are produced by such a method, there would also have to be a check for splits where the lack stratification causes fitting issues (e.g. missing or extremely rare target labels).

**Comment by Abhishek242004** [comment]
Hello @DM-Berger,

Thanks for raising this issue — I’d like to make sure I understand the intended scope correctly before attempting any contribution.
Is the main goal here to:
fail fast with a clear error or warning when stratified grouped k-folds are infeasible, or
provide guidance / recommendations (e.g. suggesting LeavePGroupsOut), or
eventually support some form of opt-in fallback behavior?

I’d be happy to help once there’s clarity on the intended direction.

**Comment by DM-Berger** [comment]
There is no main goal or direction decided on here yet, the issue is open in part for discussion and figuring out the best way to approach the issue.

**Comment by DM-Berger** [comment]
> That can be simply done by finding out the count of distinct groups for each class. If the number of splits are k then the number of groups per class should should also be at least k so that each fold might have almost the same number samples for that class.

Oh, right, another subtle point: the reason we care about grouping in df-analyze is because for certain kinds of data, e.g. data where there are multiple samples from a single individual / cohesive object (e.g. panel data, longitudinal data, embedded slices of medical imaging data from one subject, etc), then it is often trivial for classifiers to learn to predict things by simply exploiting between-individual differences (e.g. mean value) or just memorizing the classification / mean for individuals when those individuals appear in both training and test splits. This can result in huge biases, often even apparent 100% accuracies for certain kinds of data. 

That is, correctly respecting grouping is **_far_** important than perfect stratification in df-analyze (and likely in general, for most people concerned with grouping bias). So it isn't actually all that important that "each fold [...] have almost the same number samples for that class", so long as each fold has the minimum number of samples $L$ from each class, mentioned above. It also isn't even essential that all the folds be all that uniform in terms of number of samples. 

Of course, if we can generate many possible partitions of the data that respect the $L$ minimum while also not distributing any group across the train/test boundary, then obviously partitions with more uniform samples and target distributions are preferable, but grouping is the more important thing here.
